### ISSUE TITLE
Question - Package Elementary Stream of NAL units

### ISSUE THREAD

**Issue by steji113** [shaka-project/shaka-packager#224]
What would be the best way to feed NAL units into the packager to produce a manifest for a live stream? To give some background on my particular situation, I have a piece of software that ingests live video streams from IP security cameras that we would like to ultimately have accessible in the DASH format. We have access to the NAL units being received from the camera in real time before we write them out into an AVI file. In this case, would we essentially already have most of the demuxer's work done. So would we just feed samples straight into a muxer?

Thanks!

**Comment by kqyang** [comment]
We need a way to get the timestamp information, which is not contained inside the NAL units.

Does your IP camera support UDP output? If it is supported, you could use feed the UDP output to packager directly.

**Comment by steji113** [comment]
Thank you for your quick response!

We do have the timestamp information around that we can also supply. Some of our cameras do indeed support UDP, however we also support a wide range of other cameras that use different transport mechanisms for receiving the media. We support TCP and RTSP/RTP tunneled over HTTP in addition to UDP.

Is there an extension point somewhere that we could make use of in order to facilitate this use case?

**Comment by steji113** [comment]
Also, if it is helpful, we do have access to the underlying RTP media packets regardless of the transport type. Would it be possible to use something similar to the MemoryFile that can ingest the RTP packets in a circular queue?

**Comment by kqyang** [comment]
Unfortunately we don't support parsing RTP packets right now. Do you have a way of getting TS packets from all the cameras?

If not, since you have the timestamp information, the simplest option is probably to build your own version of [MediaParser](https://github.com/google/shaka-packager/blob/master/packager/media/base/media_parser.h). What you need is to construct the StreamInfo and MediaSample from the timestamp and stream data. Alternatively, you can build a RtpParser. 

Let me know if you any questions with this. I'd be happy to assist.

**Comment by steji113** [comment]
Unfortunately we cannot get TS packets as this is not too common in the security industry. =/

I will look at the MediaParser interface and see if we can implement this. We do actually have the RTP packets already parsed in our software as well, so I am hoping that will help. We make use of JRTPlib (http://jrtplib.readthedocs.io/en/stable/) to receive and parse this.

Thank you again for all of your help. I am sure I will be back with more questions. As far as I am aware, nobody has been able to achieve reliable HTML5 based H.264 video support in the physical security industry, so I am pretty sure any type of enhancements we can assist with making will be well received by the community!

**Comment by rkuroiwa** [comment]
(I could be wrong.) In order to reuse the driver program, I think both MediaParse and RTP File implementation are required. Demuxer expects a file that it can open. Changing Demuxer to get input from another source (vs. from File interface) is probably the same amount of work as writing a File implementation.

**Comment by rkuroiwa** [comment]
My suggestion might have been wrong.
tl;dr
The quickest way to support RTP using the library is to add a new MediaParser that ignores(?) the input passed to Parse(). Also comment out all the file reading logic in Demuxer class.

I think the problem is that Demuxer isn't an interface. It should be an interface which functions as a "media stream source". Also the current Demuxer class has too many non-demuxing methods (e.g. SetKeySource()) that makes it hard to swap out the implementation. (That said I think we have a plan to split out the demuxing & decrypting logic.)
IIUC, JRTPLIB takes care of both input stream & RTP parsing. Since Demuxer class assumes that input reader & parser can be separated in different modules, it's hard to use JRTPLIB without changing Demuxer.
So the quickest solution is to implement RtpMediaParser using JRTPLIB. This handles both reading the input & RTP parsing. Then comment out any File related code in the Demuxer but instead use RtpMediaParser::Parse().

**Comment by steji113** [comment]
I agree, I think this is pretty much the same conclusion I had made as well. The Demuxer does indeed introduce some implementation challenges.

Somewhat unrelated question, I am more accustomed to a MS Visual Studio development environment. What is the best way to work on this project as far as adding new classes and making them a part of the build? Any tips / pointers would be greatly appreciated! Thanks!

**Comment by rkuroiwa** [comment]
A good place to start is to add packager\media\formats\rtp\rtp.gyp and make sure packager\packager.gyp depends on it. GYP user doc is here https://gyp.gsrc.io/docs/UserDocumentation.md

If you want to use MSVS as your editor/debugger then
In cmd.exe do
set GYP_GENERATORS=ninja,msvs-ninja
set GYP_DEFINES=target_arch=ia32
(set target_arch to x64 if you're building 64bit)
python gyp_packager.py

I recommend to run ninja in cmd.exe first, somehow it has a higher chance of succeeding. For incremental builds, you should be able to build in MSVS.
Note that if you want to add a file, you have to list it in a gyp file and re-run python gyp_packager.py. Even if you are in MSVS, it uses ninja to build.

**Comment by steji113** [comment]
Thank you so much for all of the advice on this thus far, we had a few other items come up on our backlog that we had to address before now resuming on this project on our end.

Would you know what the best way is to obtain the library as a static .lib file so that we could link it to our existing application? It seems that there is quite a granular factoring of the projects, and it is not clear what the dependency graph looks like on which .lib files need to be linked against.

Thanks!

**Comment by kqyang** [comment]
We are working on exposing a standard packaging interface and including library (static/shared) in our release. It will be ready some time this month.

**Comment by steji113** [comment]
Awesome, that is great news. I believe that will be exactly what we will be looking for. Thanks again!

**Comment by steji113** [comment]
Would you happen to be able to provide any guidelines or tips on the appropriate metadata we would need in order to feed NAL units to the demuxer? We have written a custom parser that reads in a binary elementary stream format.

We start off by extracting the SPS and PPS NAL units and creating an AVCDecoderConfigurationRecord. Then we create a VideoStreamInfo with the parsed configuration record, using kCodecH264 and kNalUnitStreamWithoutParameterSetNalus (not sure if this is right). Then we subsequently loop over each frame and construct a MediaSample object from the NAL units.

A few questions:
- Are there any special considerations when encoding the H.264? We are using ffmpeg to transcode our streams on the fly.
- Should we be passing the SPS and PPS in with the first I frame?
- Should we be leaving the start codes before each NAL unit?

Thanks so much for the assistance.

**Comment by kqyang** [comment]
- Are there any special considerations when encoding the H.264? We are using ffmpeg to transcode our streams on the fly.

  No.

- Should we be passing the SPS and PPS in with the first I frame?

  kNalUnitStreamWithoutParameterSetNalus means that there is no parameter set nal units, i.e. SPS, PPS in the frame.

- Should we be leaving the start codes before each NAL unit?

  You need to convert the annexb byte stream to AVCC NAL unit stream.  You can use packager/media/codecs/h264_byte_to_unit_stream_converter.cc to do the conversion, which requires SPS/PPS in the frame. 


If you have the raw binary elementary stream in MPEG2TS, you can actually pass the raw bytes to Shaka Packager directly, then you don't need to do the operations above.

Btw, there is the interface I talked about earlier: https://github.com/google/shaka-packager/blob/master/packager/packager.h (there are still a few misc issues to clean up). Let us know if you have any comments.

**Comment by steji113** [comment]
Awesome, thanks for the input, we will give that a try with the byte to unit stream converter.

I did see the latest library pushes, this looks like what we will need, thanks for the hard work on that.
One comment: https://github.com/google/shaka-packager/blob/master/packager/packager.h#L403

An interface like this would be ideal so that our transcoding pipeline can just feed the NAL units directly. I also wonder if perhaps another function that allows passing of MediaSample's would be worth while? This would be fitting seeing as we are not actually demuxing anything (our transcoding pipeline already produces the demuxed content).

I will let you know how we fare with your suggestions.

**Comment by steji113** [comment]
So I was able to get it working finally! Thanks, very much for the suggestions, that was the trick.

Currently our implementation only reads a static file from disk and parses and packages it. Our next steps will be to attempt to use the new library and using it on a live stream.

**Comment by kqyang** [comment]
Cool!

> One comment: https://github.com/google/shaka-packager/blob/master/packager/packager.h#L403

Yes, we will address it.

> I also wonder if perhaps another function that allows passing of MediaSample's would be worth while? 

That requires exposing more internal structures (not just MediaSample) in the public interface. We are actually thinking about having a 2nd level interface which allows more controls of the library, but it won't happen in the near term.

**Comment by steji113** [comment]
I think the first portion will be sufficient for our needs. Also, another request, what do you think about exposing the ability to set a specific parser implementation via a public method on the Demuxer class? That way, in our particular use case, we could skip the portion where the Demuxer tries to intelligently determine the container format. Since we are just using an elementary stream, we have a need to inject some additional information, such as the timescale and the FPS for us to be able to appropriately timestamp the frames.

**Comment by kqyang** [comment]
> Also, another request, what do you think about exposing the ability to set a specific parser implementation via a public method on the Demuxer class?

SGTM. We can move the creation of the parser to a MediaParserFactory class and allow injection of MediaParserFactory. If you have something already, feel free to submit a pull request.

**Comment by steji113** [comment]
I don't have anything in progress on that front yet, but I would be willing to take a stab at it once I begin work on the library version of our implementation.

**Comment by steji113** [comment]
So I attempted to link the libpackager static library to our code today and ran into a number of linker errors. I am not sure if this is particular to any of our project settings or if it is currently an issue with the dependencies for libpackager. Please see this gist: **https://gist.github.com/steji113/2e0a6ef2ca250eae55e274462fb008de**

I will spend a bit more time researching and seeing if the issue is on our side. I was not sure if you have tried linking with any external sample applications yet.

**Comment by steji113** [comment]
I am wondering if it has anything to do with this: https://chromium.googlesource.com/chromium/src/+/lkcr/docs/component_build.md and whether or not we just need to tweak some settings in the packager.gyp.

**Comment by kqyang** [comment]
https://github.com/google/shaka-packager/blob/master/packager/packager_test.cc uses libpackager and does not have internal dependencies except on gtest and glogging. 

I don't think your problem is related to https://chromium.googlesource.com/chromium/src/+/lkcr/docs/component_build.md. Do you have a minimal code snippet that can duplicate the issue?

**Comment by steji113** [comment]
We can put a sample together and see if it has something to do with our project configuration. I just wanted to rule out any low hanging fruit.

**Comment by rkuroiwa** [comment]
I think the static build (libpackager.a) doesn't contain libbase.a. Try linking against libpackager.so and it should contain a copy of base.

kqyang: There might be a GYP flag to tell it to pull everything from its dependencies into the packager target.

**Comment by steji113** [comment]
[ShakaPackagerTest.zip](https://github.com/google/shaka-packager/files/1089469/ShakaPackagerTest.zip)

I have created a minimum reproduction. We are in a Windows environment, too.

**Comment by steji113** [comment]
@rkuroiwa This is kind of what I was thinking as well. It seems that a lot of the unresolved symbols are from the base classes.

**Comment by kqyang** [comment]
I tried it. It actually works with static libpackager. A minor tweak is needed to make it work with dynamic libpackager: https://gist.github.com/kqyang/4031f2ee70b16879de81158da96f4df7.

@rkuroiwa GYP pulls all "static" targets automatically actually, and libpackager.a does include libbase.a.

@steji113 Did you touch is_component_build flag? It should not be modified and should stay as false.

**Comment by steji113** [comment]
@kqyang So I wonder if this is the source of my difficulties. How do you build only the static library from ninja / gyp? I don't believe I have touched the is_component_build flag as I am not sure where that is. I have just been running this:
```
set GYP_GENERATORS=ninja,msvs-ninja
set GYP_DEFINES=target_arch=x64
python gyp_packager.py
ninja -C out/Debug_x64
```

Ultimately, there is only a libpackager.lib in out/Debug_x64/obj. I never get one in out/Debug_x64/lib as I had expected.

Also, if I look at the packager.ninja, I see this:
build packager.exe packager.exe.pdb: link_embed $
    obj\app\packager.crypto_flags.obj $
    obj\app\packager.fixed_key_encryption_flags.obj $
    obj\app\packager.gflags_hex_bytes.obj obj\app\packager.hls_flags.obj $
    obj\app\packager.mpd_flags.obj obj\app\packager.muxer_flags.obj $
    obj\app\packager.packager_main.obj $
    obj\app\packager.playready_key_encryption_flags.obj $
    obj\app\packager.retired_flags.obj $
    obj\app\packager.stream_descriptor.obj $
    obj\app\packager.validate_flag.obj obj\app\packager.vlog_flags.obj $
    obj\app\packager.widevine_encryption_flags.obj obj\base\base.lib $
    obj\libpackager.lib obj\media\file\file.lib $
    obj\third_party\gflags\gflags.lib obj\base\allocator\allocator.lib $
    obj\base\base_static.lib obj\third_party\modp_b64\modp_b64.lib $
    obj\base\third_party\dynamic_annotations\dynamic_annotations.lib $
    obj\hls\hls_builder.lib obj\media\base\media_base.lib $
    obj\media\base\widevine_pssh_data_proto.lib $
    obj\third_party\protobuf\protobuf_full_do_not_use.lib $
    obj\third_party\boringssl\boringssl.lib $
    obj\third_party\boringssl\boringssl_asm.lib $
    obj\third_party\curl\libcurl.lib obj\third_party\libxml\libxml2.lib $
    obj\third_party\icu\icuuc.lib obj\third_party\zlib\zlib.lib $
    obj\third_party\zlib\zlib_x86_simd.lib obj\version\version.lib $
    obj\mpd\media_info_proto.lib obj\media\codecs\codecs.lib $
    obj\media\chunking\chunking.lib obj\media\demuxer\demuxer.lib $
    obj\media\formats\mp2t\mp2t.lib obj\media\crypto\crypto.lib $
    obj\media\formats\mp4\mp4.lib obj\media\event\media_event.lib $
    obj\media\formats\mpeg\mpeg.lib obj\media\formats\webm\webm.lib $
    obj\third_party\libwebm\mkvmuxer.lib obj\media\formats\cv\cv.lib $
    obj\media\formats\webvtt\webvtt.lib obj\media\formats\wvm\wvm.lib $
    obj\media\trick_play\trick_play.lib obj\mpd\mpd_builder.lib | $
    ..\..\packager\build\win\compatibility.manifest $
    gen\base\allocator\features.h gen\base\debug\debugging_flags.h $
    gen\base\generated_build_date.h gen\base\win\base_features.h icudtl.dat $
    obj\base\trace_event\etw_manifest\etw_manifest.actions_rules_copies.stamp $
    obj\packager.packager.exe.generated.manifest || $
    obj\packager.actions_depends.stamp

Which leads me to believe that it is breaking out everything into separate libraries.

**Comment by kqyang** [comment]
@steji113 @rkuroiwa You are right. The targets libpackager depends on are not built into libpackager.a. I made a mistake earlier. My gyp test target works does not mean that libpackager.a includes these dependencies. In fact, the dependencies are just propagated to my test target through gyp. We'll need the below setting to make it work:

     'target_name': 'libpackager',
     'type': '<(libpackager_type)',
     'standalone_static_library': 1,
     'sources': [

I'll fix it tomorrow.

**Comment by steji113** [comment]
@kqyang awesome, thanks! I just wanted to make sure I wasn't going crazy here.

**Comment by kqyang** [comment]
Fixed in https://github.com/google/shaka-packager/commit/a678948db4df87b2c4d5c140e8b61c267388fbbf

**Comment by steji113** [comment]
Awesome, will check it out and let you know if I find anything else. Thanks!

**Comment by steji113** [comment]
@kqyang So a new issue has turned up now. I guess my large Windows based application that I am trying to link against is compiled with MultiByte character support and it seems the libpackager uses Unicode. Do you know if you guys have any plans for supporting both?

**Comment by kqyang** [comment]
@steji113 No, we don't have plan to support MBCS. You'll have to do the conversions outside of libpackager, unfortunately.

**Comment by steji113** [comment]
@kqyang I am back working on this now, I had to jump on a few other things for a while.

In order to avoid the libpackager character set issue I referred to above, I was thinking of using libpackager as a DLL instead. However, I seem to be running into issues compiling as a "shared_library." I am compiling with VS 2015 on x64 on Windows.

```
set GYP_DEFINES=target_arch=x64 libpackager_type=shared_library
gclient runhooks
cd src
ninja -C out/Debug_x64
```
I get errors such as:
```
[1/47] CXX obj\status.status.obj
FAILED: obj/status.status.obj 
ninja -t msvc -e environment.x64 -- "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\cl.exe" /nologo /showIncludes /FC @obj\status.status.obj.rsp /c ..\..\packager\status.cc /Foobj\status.status.obj /Fdobj\status.cc.pdb 
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2143: syntax error: missing ')' before '('
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2079: 'visibility' uses undefined class 'shaka::__attribute__'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2059: syntax error: ')'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2440: 'initializing': cannot convert from 'const char [8]' to 'int'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): note: There is no context in which this conversion is possible
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C2146: syntax error: missing ';' before identifier 'Status'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(110): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(111): error C2059: syntax error: 'public'
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(113): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
c:\users\stephen.daudell\documents\code\shaka_me\src\packager\status.h(113): error C2365: 'shaka::Status': redefinition; previous definition was 'data variable'
...MORE ERRORS HERE
```

I noticed #227, I wonder if building as a shared library might be broken again? Building as a standalone static library works fine for me.

Thanks!

**Comment by kqyang** [comment]
Yes, shared library build is broken on Windows right now. Feel free to open a new issue. We don't have resource to work on the fix right now - you are welcome to try :)

As for the above error, it can be fixed by replacing OS_WIN with WIN32. I suspect that there are other errors..

**Comment by steji113** [comment]
OK, thank you for the confirmation. I will open a bug for Windows to track that.
